# Softwares used in the snv_indels module

## [bcbio_variation_recall_ensemble](https://github.com/bcbio/bcbio.variation.recall)
Ensemble variants in the `.vcf`-file format from different callers in one `.vcf`-file.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bcbio_variation_recall__bcbio_variation_recall_ensemble#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bcbio_variation_recall__bcbio_variation_recall_ensemble#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__bcbio_variation_recall_ensemble#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__bcbio_variation_recall_ensemble#

---

## [bcftools_concat](https://samtools.github.io/bcftools/bcftools.html)
Concatenate `.vcf`-files from different chromosomes using bcftools concat.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bcftools__bcftools_concat#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bcftools__bcftools_concat#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__bcftools_concat#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__bcftools_concat#

---

## [bcftools_sort](https://samtools.github.io/bcftools/bcftools.html)
Sort a `.vcf`-file using bcfools sort.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bcftools__bcftools_sort#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bcftools__bcftools_sort#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__bcftools_sort#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__bcftools_sort#

---

## [bcftools_view](https://samtools.github.io/bcftools/bcftools.html)
Convert bcftools `.bcf` binary file format to a `.vcf`-file using bcftools.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bcftools__bcftools_view#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bcftools__bcftools_view#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__bcftools_view#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__bcftools_view#

---

## [bed_split](https://linux.die.net/man/1/awk)
Split a bed file into chromosomes using an awk command going by the first column (chromosome).

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bed_split__bed_split#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bed_split__bed_split#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__bed_split#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__bed_split#

---

## [bgzip](http://www.htslib.org/doc/bgzip.html)
Compress a `.vcf`-file using bgzip.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bgzip__bgzip#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bgzip__bgzip#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__bgzip#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__bgzip#

---

## [deeptrio_call_variants](https://github.com/google/deepvariant)
Step 2 of 3 in the calling of SNVs and INDELs using deeptrio.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__deeptrio__deeptrio_call_variants#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__deeptrio__deeptrio_call_variants#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__deeptrio_call_variants#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__deeptrio_call_variants#

---

## [deeptrio_make_examples](https://github.com/google/deepvariant)
Step 1 of 3 in the calling of SNVs and INDELs using deeptrio.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__deeptrio__deeptrio_make_examples#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__deeptrio__deeptrio_make_examples#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__deeptrio_make_examples#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__deeptrio_make_examples#

---

## [deeptrio_postprocess_variants](https://github.com/google/deepvariant)
Last step in the calling of SNVs and INDELs using deeptrio.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__deeptrio__deeptrio_postprocess_variants#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__deeptrio__deeptrio_postprocess_variants#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__deeptrio_postprocess_variants#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__deeptrio_postprocess_variants#

---

## fix_af
Fixes missing AF in the format field of a vcf-file as this is not generated by all callers.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__fix_af__fix_af#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__fix_af__fix_af#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__fix_af#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__fix_af#

---

## freebayes
Fixes missing AF in the format field of a vcf-file as this is not generated by all callers.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__freebayes__freebayes#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__freebayes__freebayes#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__freebayes#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__freebayes#

---

## gatk_mutect2
Step 1 of 4 of Mutect2 variant calling generating chromosome split unfiltered calls as well as variant statistic files (used for variant filtering) and INDEL-bam files.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__gatk__gatk_mutect2#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__gatk__gatk_mutect2#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__gatk_mutect2#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__gatk_mutect2#

---

## gatk_mutect2_gvcf
Mutect2 is used to generate a genome vcf file containing allele information for all positions in the bed file.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__gatk__gatk_mutect2_gvcf#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__gatk__gatk_mutect2_gvcf#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__gatk_mutect2_gvcf#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__gatk_mutect2_gvcf#

---

## gatk_mutect2_filter
Step 3 of 4 of Mutect2 variant calling filtering the called variants using the merged statistics file.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__gatk__gatk_mutect2_filter#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__gatk__gatk_mutect2_filter#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__gatk_mutect2_filter#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__gatk_mutect2_filter#

---

## gatk_mutect2_merge_stats
Step 2 of 4 of Mutect2 variant calling merging the statistics file.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__gatk__gatk_mutect2_merge_stats#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__gatk__gatk_mutect2_merge_stats#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__gatk_mutect2_merge_stats#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__gatk_mutect2_merge_stats#

---